### ISSUE TITLE
Add `expand` tensor operation similar to PyTorch

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn#1412]
### Summary

I am proposing the addition of an `expand` tensor operation to the Burn deep learning framework, mirroring the functionality available in PyTorch. This operation is invaluable for broadcasting operations on tensors of different shapes, making certain types of tensor manipulations more intuitive and efficient.

### Motivation

In PyTorch, the `expand` method allows users to expand the size of an existing tensor without copying data. This is particularly useful for performing operations that require tensors to be of the same shape, enabling more memory-efficient computations by avoiding the creation of large intermediate tensors. Its inclusion in Burn would greatly enhance the framework's tensor manipulation capabilities, aligning it more closely with PyTorch's feature set and making it more appealing to users transitioning between frameworks or leveraging both for different aspects of their work.

### Technical Details

The `expand` function in PyTorch allows a tensor to be "viewed" with a larger size, with the new dimensions being broadcastable according to the broadcasting rules of tensors. The operation does not allocate new memory for the expanded tensor but provides a new view on the existing tensor, which is a very efficient way to perform many element-wise operations.

I propose that the API for the `expand` operation in Burn should closely mirror that of PyTorch for ease of use and familiarity. 

### Use Cases

This feature is crucial for operations where matching tensor dimensions is necessary without the need to duplicate data. For example, when adding a vector to each row of a matrix, or applying a mask across multiple dimensions of a tensor where the mask is initially only defined in a subset of those dimensions.

### Example

In PyTorch, the usage of `expand` is as follows:

```python
import torch
x = torch.tensor([1, 2, 3])
x_expanded = x.expand(3, 3)
# x_expanded is now [[1, 2, 3], [1, 2, 3], [1, 2, 3]] without copying data
```

A similar functionality in Burn would enable users to perform such operations with the same ease and efficiency.

### Conclusion

The addition of an `expand` tensor operation to Burn could significantly improve its usability and efficiency for a wide range of tasks, making it a more competitive and versatile tool in the deep learning ecosystem. I believe this feature would be a valuable asset to the framework and its users.

**Comment by antimora** [comment]
From Discord discussion:

> [4:15 AM]Zermelo Fraenkel: Just to mention that on the candle side, we handle expand by just setting some stride to 0s, so no copy is involved in the process (until an op is actually performed on the tensor), I guess the same is done in other frameworks.

> [7:47 AM]Nathaniel: I though exactly of that, but then, when is it preferable to use repeat?

> [9:26 AM]Zermelo Fraenkel: I don't think there is a very good case for doing a "concretised" repeat, you might need it if the next op is a kernel that requires an actual contiguous tensor e.g. matmul for some candle backends but it's more of an edge case.  repeat is just a shortcut for .expand().contiguous() and this seems a bit annoying to have such a subtle difference even if it's inherited from pytorch https://pytorch.org/docs/stable/generated/torch.Tensor.repeat.html
